crates/re_query/src/archetype_view.rs

```diff
@@ -322,24 +322,60 @@ impl<A: Archetype> ArchetypeView<A> {
         let component = self.components.get(&C::name());
 
         if let Some(component) = component {
-            let primary_instance_key_iter = self.iter_instance_keys();
+            // NOTE: Autogenerated instance keys are interned: if two or more rows in the store
+            // share the same keys, then they will literally point to the same memory address.
+            // Therefore we can compare those addresses, and early out if they match.
+            let primary_instance_key_buffer =
+                self.required_comp().instance_keys.as_arrow_ref() as *const dyn Array;
+            let component_instance_key_buffer =
+                component.instance_keys.as_arrow_ref() as *const dyn Array;
+
+            // TODO(rust-lang/rust#81513): We want to compare just the data part of these fat pointers,
+            // because comparing the vtable pointers, while defined, can lead to unstable results.
+            //
+            // We'd like to use `std::ptr::to_raw_parts` for that (or any of the myriad of tools
+            // for dealing with smart pointers available in the stdlib)… but everything that
+            // relates to fat pointers is nightly only.
+            // So, for now, we do it the old way.
+            if std::ptr::eq(
+                primary_instance_key_buffer.cast::<u8>(),
+                component_instance_key_buffer.cast::<u8>(),
```

Comment:
We should also check that the lengths are equal, otherwise we will get breakage when we switch to a smarter prefix interning (0..10 and 0..100 being different slices of the same buffer)

Reply:
Your comment actually made me realize that we can avoid all this dark magic by working at the `DataCell` layer rather than raw Arrow & trait objects.

https://github.com/rerun-io/rerun/pull/3377/commits/f40d3a3240baa22741fbbc3859c3556fd56e1614 removes the black magic.